advisories/2024-09-25_cometbft-bls12381-key-validation.md

```diff
@@ -0,0 +1,126 @@
+# Missing BLS12-381 Public Key Validation
+
+## Description
+
+CometBFT omitted the subgroup check on the public keys of the BLS signature scheme. The subgroup check ensures that the public key is a point on the subgroup of prime order `r` of the curve. The omission would not only have allowed for computations in arbitrary groups but, assuming a Cosmos application using the BLS12-381 keys for a key-exchange protocol, an attacker whose public key was not validated could have induced the other participant in the protocol to inadvertently perform computations with her secret key on the invalid public key, an operation which is known to leak data about the secret key [1].
+
+CometBFT latest available release was, at the time of this writing, v1.0.0-rc1.0.20240805092115-3b2c5d9e1843, which introduced support for BLS signatures over the BLS12-381 curve. In this release, instantiation of BLS public keys from byte arrays is done via `github.com/cosmos/crypto v0.0.0-20240309083813-82ed2537802e`, a clone of the Prysm's wrapper over BLST, as it can be seen at: https://github.com/cometbft/cometbft/blob/86da0027d878707365c16b124b77892ca5212fe1/crypto/bls12381/key_bls12381.go#L127 The `PublicKeyFromBytes` function performs the G1 subgroup and infinity check with a call to BLST's `KeyValidate`: https://github.com/cosmos/crypto/blob/bb8c5deb91b3a722e145c4d9c6d06c6158d23dfe/curves/bls12381/pubkey.go#L64
```

Comment:
Links like these should be made hyperlinks to improve readibility. For example:

a clone of the Prysm's warpper over BLST, available [here](https://github.com/cometbft/cometbft/blob/86da0027d878707365c16b124b77892ca5212fe1/crypto/bls12381/key_bls12381.go#L127)

Reply:
```suggestion
At the time of writing, CometBFT's latest available release was `v1.0.0-rc1.0.20240805092115-3b2c5d9e1843`. This release introduced support for BLS signatures over the BLS12-381 curve. In this release, instantiation of BLS public keys from byte arrays was done using `github.com/cosmos/crypto v0.0.0-20240309083813-82ed2537802e`, a clone of the Prysm's wrapper over BLST, seen [here](https://github.com/cometbft/cometbft/blob/86da0027d878707365c16b124b77892ca5212fe1/crypto/bls12381/key_bls12381.go#L127). The `PublicKeyFromBytes` function performs the G1 subgroup and infinity check with a call to BLST's [`KeyValidate`](https://github.com/cosmos/crypto/blob/bb8c5deb91b3a722e145c4d9c6d06c6158d23dfe/curves/bls12381/pubkey.go#L64)
```